.pre-commit-config.yaml

```diff
@@ -0,0 +1,37 @@
+# See https://pre-commit.com for more information
+# See https://pre-commit.com/hooks.html for more hooks
+repos:
+-   repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: 'v4.0.1'
+    hooks:
+        # Prevent giant files from being committed.
+    -   id: check-added-large-files
+        # Check whether files parse as valid Python.
+    -   id: check-ast
+        # Check for file name conflicts on case-insensitive filesytems.
+    -   id: check-case-conflict
+        # Check for files that contain merge conflict strings.
+    -   id: check-merge-conflict
+        # Check for debugger imports and py37+ `breakpoint()` calls in Python source.
+    -   id: debug-statements
+        # Don't commit to master branch.
+    -   id: no-commit-to-branch
+-   repo: https://github.com/psf/black
+    rev: '21.5b2'
+    hooks:
+    -   id: black
+        # Force black to run on whole repo, using settings from pyproject.toml
+        pass_filenames: false
+        args: [--config=./pyproject.toml, .]
+-   repo: https://github.com/PyCQA/flake8
+    rev: '3.9.2'
+    hooks:
+        # Run flake8.
+    -   id: flake8
+        args: [--config=./.flake8]
+-   repo: https://github.com/pycqa/isort
+    rev: '5.8.0'
```

Comment:
It's a bit of a shame that the way this config works is that you have to pin to a specific tag, rather than just get the latest.

Do we plan to just regularly update these revs? I'm just wondering what the process of that will be (and will we remember!)

Reply:
So I don't think this is so bad...

Basically, we've now have a `pre-commit.ci` bot automatically ensuring that the `.pre-commit-config` entries are up-to-date. The bot will automatically create a branch and pull-request to update `.pre-commit-config.yaml` with the latest versions, as they happen.

Also, for `cirrus-ci` it `pip` installs the `latest` versions of `black`, `flake8` and `isort`, and this should align with the versions in the `.pre-commit-config.yaml`... so it's all kinda joined up.

We could add extra boiler plate code to ensure that the versions installed by `cirrus-ci` is scraped from the `.pre-commit-config.yaml`, but I don't think the effort is worth the benefit .

So, all in all, it's pretty automated and hinges on `pre-commit.ci` doing all the hard work.

Does this seem reasonable?